qcodes_contrib_drivers/tests/test_Spectrum_M4i.py

```diff
@@ -0,0 +1,21 @@
+import unittest
+from unittest.mock import MagicMock
+from unittest.mock import patch
+
+
+class TestM2j(unittest.TestCase):
+
+    @staticmethod
+    def test_M4i_wait_ready():
+
+        mock_pyspcm_module = MagicMock(name='pyspcm')
+
+        with patch.dict('sys.modules', pyspcm=mock_pyspcm_module):
+            import qcodes.instrument_drivers.Spectrum.M4i
+            m4i = qcodes.instrument_drivers.Spectrum.M4i.M4i('test_m4i_instrument')
+
+            m4i.sample_rate()
+            mock_pyspcm_module.int32.assert_called()
+            m4i.wait_ready()
+            mock_pyspcm_module.spcm_dwSetParam_i32.assert_called()
+            m4i.close()
```

Comment:
Please move setting up the mock and instantiating the instrument to setup method and closing the instrument (preferably via its class `M4i.cloase_all()`) to teardown method of the test case class.

Also, I recommend using pytest instead of unittest.

Reply:
@astafan8 How can I move the instrument instantiating to the `setUp`? The instrument needs a patch of the `pyspcm` module which I now do with a `with patch.dict(...)`. I would like to keep the patch atomic and not spread it out over a setUp and tearDown.

Reply:
> I would like to keep the patch atomic and not spread it out over a setUp and tearDown

in this case, ok, don't move the instatiation of the instrument to setUp but then right after it's initiated please add an "automatic closing the instrument" using [`self.addCleanup` method](https://docs.python.org/3/library/unittest.html#unittest.TestCase.addCleanup) (for example `self.addCleanup(M4i.close_all)`)